optimize: fix `effect_free` refinement in post-opt dataflow analysis

aviatesk opened the pull request:
Currently we never refine this information though.

@nanosoldier `runbenchmarks("inference", vs=":master")`

aviatesk committed:
optimize: fix `effect_free` refinement in post-opt dataflow analysis

Currently we never refine this information though.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51185) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/14cfe87_vs_1829cee/report.md).

vtjnash commented:
Anyway to demonstrate this with a test?

aviatesk commented:
I don't think there are easy way to kick this refinement currently, because the optimizer doesn't derive new information about `effect_free`-ness, i.e. if a method is not `:effect_free`, then the optimized body has at least one statement without `IR_FLAG_EFFECT_FREE`. We need to integrate EA into the optimizer to refine `:effect_free`.

vtjnash commented:
Alright, just waiting for Keno's review on this and #51211 then

Keno commented:
Currently, I think the only place where effect_free would get refined is if the effect is in a branch discovered to be dead by mutable SROA propagation, since we don't handle that in inference. Maybe try writing a test case around that?

aviatesk commented:
Yeah, I tried it, but it didn't work because currently we don't mark `IR_FLAG_EFFECT_FREE` for `ReturnNode`. I tried the following diff too, but it seems to cause another issue:
```diff
diff --git a/base/compiler/optimize.jl b/base/compiler/optimize.jl
index c4f8874eab..90794783cc 100644
--- a/base/compiler/optimize.jl
+++ b/base/compiler/optimize.jl
@@ -272,7 +272,7 @@ function stmt_effect_flags(𝕃ₒ::AbstractLattice, @nospecialize(stmt), @nospe
     # TODO: We're duplicating analysis from inference here.
     isa(stmt, PiNode) && return (true, true, true)
     isa(stmt, PhiNode) && return (true, true, true)
-    isa(stmt, ReturnNode) && return (true, false, true)
+    isa(stmt, ReturnNode) && return (true, isdefined(stmt, :val), isdefined(stmt, :val))
     isa(stmt, GotoNode) && return (true, false, true)
     isa(stmt, GotoIfNot) && return (true, false, ⊑(𝕃ₒ, argextype(stmt.cond, src), Bool))
     if isa(stmt, GlobalRef)
diff --git a/base/compiler/ssair/ir.jl b/base/compiler/ssair/ir.jl
index fa857b4f70..2f638b3914 100644
--- a/base/compiler/ssair/ir.jl
+++ b/base/compiler/ssair/ir.jl
@@ -1730,6 +1730,7 @@ function maybe_erase_unused!(callback::Function, compact::IncrementalCompact, id
     inst = idx ≤ nresult ? compact.result[idx] : compact.new_new_nodes.stmts[idx-nresult]
     stmt = inst[:stmt]
     stmt === nothing && return false
+    stmt isa ReturnNode && return false
     inst[:type] === Bottom && return false
     effect_free = (inst[:flag] & (IR_FLAG_EFFECT_FREE | IR_FLAG_NOTHROW)) == IR_FLAG_EFFECT_FREE | IR_FLAG_NOTHROW
     effect_free || return false
```

Keno commented:
Yeah, I ran into the same thing. I don't know if we should really set IR_FLAG_EFFECT_FREE on the control flow nodes. Maybe just ignore them in the ipo dataflow analysis?

aviatesk committed:
post-opt analysis: ignore control flow nodes when computing `all_effect_free`

Also ignores `nothing` statement, because it can be from control flow
nodes that have `IR_FLAG_EFFECT_FREE` set.

aviatesk committed:
post-opt-analysis: add test case for `:effect_free` refinement

aviatesk commented:
Okay thanks for you tips, I think I got it working. It needed a tweak on CFG transformation that adds `IR_FLAG_EFFECT_FREE` on statement that replaced `GotoIfNot` with constant condition too.

Keno commented on base/compiler/optimize.jl:
```diff
@@ -570,9 +570,14 @@ function ipo_dataflow_analysis!(interp::AbstractInterpreter, ir::IRCode, result:
+    function scan_non_dataflow_flags!(inst::Instruction, @nospecialize(stmt=inst[:stmt]))
         flag = inst[:flag]
-        all_effect_free &= (flag & IR_FLAG_EFFECT_FREE) != 0
+        if !isterminator(stmt)
```
Let's add `&& stmt !== nothing` so that statement deletion is automatically considered effect free. Then we don't need this in the other place.

aviatesk replied:
I thought https://github.com/JuliaLang/julia/pull/51185/commits/b0f30b10aac134402fb7d9896562911bd2eac9fd is a more general solution because we may observe the IR flag in other places too. But I think only this postopt analysis and the statement deletion observes it, and since the latter already has a special case for `nothing`, we may want to have it here too.
```suggestion
        if !isterminator(stmt) && stmt !== nothing
```

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51185#issuecomment-1709569846) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/280080c_vs_aeae142/report.md).

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51185#issuecomment-1709818521) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/5b54ac9_vs_1305f40/report.md).

aviatesk commented:
@nanosoldier `runbenchmarks("inference", vs=":master")`

Keno commented:
Not sure I entirely understand the 5% regression, but I guess it's possible that fixing the effect_free bug causes it to take longer to bail out (because it was incorrectly bailing out early before). Regardless, I'm gonna go ahead and merge this since I'd like to do some further work on top.

aviatesk commented:
I'd say the 5% regression in time ratio is in a range of noise. There are no increased allocations at least.
